sql/enumtype.go

```diff
@@ -0,0 +1,264 @@
+package sql
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"gopkg.in/src-d/go-errors.v1"
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/proto/query"
+)
+
+var ErrConvertingToEnum = errors.NewKind("value %v is not valid for this Enum")
+
+// Comments with three slashes were taken directly from the linked documentation.
+
+// Represents the ENUM type.
+// https://dev.mysql.com/doc/refman/8.0/en/enum.html
+type EnumType interface {
+	Type
+	At(index int) (string, bool)
+	CharacterSet() CharacterSet
+	Collation() Collation
+	ConvertToIndex(v interface{}) (int, error)
+	IndexOf(v string) int
+}
+
+type enumType struct{
+	collation Collation
+	valToIndex map[string]int
+	indexToVal []string
+}
+
+// CreateEnumType creates a EnumType.
+func CreateEnumType(values []string, collation Collation) (EnumType, error) {
+	if len(values) == 0 {
+		return nil, fmt.Errorf("number of values may not be zero")
+	}
+	/// An ENUM column can have a maximum of 65,535 distinct elements.
+	if len(values) > 65535 {
+		return nil, fmt.Errorf("number of values is too large")
+	}
+	valToIndex := make(map[string]int)
+	for i, value := range values {
+		if collation != Collation_binary {
+			/// Trailing spaces are automatically deleted from ENUM member values in the table definition when a table is created.
+			value = strings.TrimRight(value, " ")
+		}
+		values[i] = value
```

Comment:
Shouldn't this be i+1?

Reply:
I saw later in the review how you're handling this in the At() method, but this would be more straightforwardly expressed as either a map or an array with no 0 index value.